k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java

```diff
@@ -1652,6 +1654,104 @@ private <MSG extends Message> void updateStatus(
+            @Override
+            public void run() {
+                try {
+                    tryToDecryptMessagesThatCouldNotDecryptBeforeSynchronous();
+                } catch (MessagingException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        });
+
+    }
+
+    private synchronized void tryToDecryptMessagesThatCouldNotDecryptBeforeSynchronous() throws MessagingException {
+        for (Account account : preferences.getAvailableAccounts()) {
+            List<LocalFolder> localFolders = account.getLocalStore().getPersonalNamespaces(false);
+            for (LocalFolder localFolder : localFolders) {
+                tryToDecryptMessagesThatCouldNotDecryptBeforeSynchronous(account, localFolder);
+            }
+        }
+    }
+
+    private void tryToDecryptMessagesThatCouldNotDecryptBeforeSynchronous(
+            Account account,
+            LocalFolder localFolder
+    ) throws MessagingException {
+        Folder<? extends Message> remoteFolder = null;
+        try {
+            if (shouldTryToDecryptEncryptedMessagesForFolder(localFolder)) {
+                final RemoteStore remoteStore = account.getRemoteStore();
+                localFolder.open(Folder.OPEN_MODE_RW);
+                remoteFolder = remoteStore.getFolder(localFolder.getName());
+                remoteFolder.open(Folder.OPEN_MODE_RO);
+                String[] idsToProcess = preferences.getStorage()
+                        .getCouldNotDecryptMessages().toArray(new String[0]);
+
+                for (String couldNotDecryptMessageId : idsToProcess) {
+                    tryToDecryptMessageThatCouldNotDecrypt(
+                            account,
+                            localFolder,
+                            remoteFolder,
+                            couldNotDecryptMessageId
+                    );
+                }
+            }
+        } finally {
+            closeFolder(localFolder);
+            closeFolder(remoteFolder);
+        }
+    }
+
+    private static boolean shouldTryToDecryptEncryptedMessagesForFolder(LocalFolder localFolder) { // TODO: 18/1/24 review this condition
+        return localFolder.getDisplayClass().equals(Folder.FolderClass.FIRST_CLASS)
+                && !localFolder.getName().equals(Account.OUTBOX);
+    }
+
+    private void tryToDecryptMessageThatCouldNotDecrypt(
+            Account account,
+            LocalFolder localFolder,
+            Folder<? extends Message> remoteFolder,
+            String couldNotDecryptMessageId
+    ) throws MessagingException {
+        String newUid = remoteFolder.getUidFromMessageId(couldNotDecryptMessageId); // get real uid just in case it changed
+        if (newUid == null) return;
+        String folderName = localFolder.getName();
+        LocalMessage couldNotDecryptMessage = loadMessageWithoutMarkingRead(account, folderName, newUid);
+        if (couldNotDecryptMessage != null) {
```

Comment:
this var name is a bit confusing, when I read the code first I thought you would enter this if condition if you could not decrypt a message, but that's basically instead a check that the message that you failed at decrypting actually exists, before trying again. Maybe we could call it messageToDecrypt o messageToReTryDecrypting